File: riveredge-frontend/src/apps/master-data/services/drawingDistribution.ts
/**
 * 图档发放单 API
 */

import { api } from '../../../services/api';

export type DrawingDistributionStatus = 'Draft' | 'Pending' | 'Issued' | 'Recalled';

export interface DrawingDistributionLine {
  id: number;
  drawingId: number;
  drawingUuid: string;
  drawingCode: string;
  drawingName: string;
  drawingRevision: string;
}

export interface DrawingDistribution {
  id: number;
  uuid: string;
  tenantId: number;
  code: string;
  name: string;
  status: DrawingDistributionStatus;
  remark?: string;
  issuedAt?: string;
  issuedByName?: string;
  recalledAt?: string;
  recalledByName?: string;
  recallReason?: string;
  createdByName?: string;
  updatedByName?: string;
  createdAt: string;
  updatedAt: string;
  lines: DrawingDistributionLine[];
}

export interface DrawingDistributionListResponse {
  data: DrawingDistribution[];
  total: number;
}

export interface DrawingDistributionPolicy {
  isEnabled: boolean;
}

const BASE = '/apps/master-data/process/drawing-distributions';

export const drawingDistributionApi = {
  list: async (params?: {
    skip?: number;
    limit?: number;
    status?: string;
    keyword?: string;
  }): Promise<DrawingDistributionListResponse> => {
    return api.get<DrawingDistributionListResponse>(BASE, { params });
  },

  get: async (uuid: string): Promise<DrawingDistribution> => {
    return api.get<DrawingDistribution>(`${BASE}/${uuid}`);
  },

  create: async (data: {
    code?: string;
    name: string;
    remark?: string;
    lines: Array<{ drawingUuid: string }>;
  }): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(BASE, data);
  },

  update: async (
    uuid: string,
    data: { name?: string; remark?: string; lines?: Array<{ drawingUuid: string }> },
  ): Promise<DrawingDistribution> => {
    return api.put<DrawingDistribution>(`${BASE}/${uuid}`, data);
  },

  delete: async (uuid: string): Promise<void> => {
    return api.delete(`${BASE}/${uuid}`);
  },

  submit: async (uuid: string): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(`${BASE}/${uuid}/submit`, {});
  },

  approve: async (uuid: string): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(`${BASE}/${uuid}/approve`, {});
  },

  reject: async (uuid: string): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(`${BASE}/${uuid}/reject`, {});
  },

  revoke: async (uuid: string): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(`${BASE}/${uuid}/revoke`, {});
  },

  recall: async (uuid: string, reason?: string): Promise<DrawingDistribution> => {
    return api.post<DrawingDistribution>(`${BASE}/${uuid}/recall`, { reason });
  },

  getPolicy: async (): Promise<DrawingDistributionPolicy> => {
    return api.get<DrawingDistributionPolicy>(`${BASE}/policy`);
  },

  updatePolicy: async (isEnabled: boolean): Promise<DrawingDistributionPolicy> => {
    return api.put<DrawingDistributionPolicy>(`${BASE}/policy`, { isEnabled });
  },
};

File: riveredge-backend/src/apps/master_data/models/drawing_distribution.py
"""图档发放单与发放控制策略"""

from tortoise import fields

from core.models.base import BaseModel


class DrawingDistributionPolicy(BaseModel):
    class Meta:
        table = "apps_master_data_drawing_distribution_policies"
        table_description = "基础数据管理 - 图档发放控制"
        indexes = [("tenant_id",)]

    id = fields.IntField(pk=True, description="主键ID")
    is_enabled = fields.BooleanField(default=False, description="启用后车间只读已发放未收回版")
    updated_by = fields.IntField(null=True, description="更新人ID")
    updated_by_name = fields.CharField(max_length=100, null=True, description="更新人姓名")


class DrawingDistribution(BaseModel):
    class Meta:
        table = "apps_master_data_drawing_distributions"
        table_description = "基础数据管理 - 图档发放单"
        indexes = [
            ("tenant_id",),
            ("code",),
            ("status",),
            ("created_at",),
        ]

    id = fields.IntField(pk=True, description="主键ID")
    code = fields.CharField(max_length=50, description="发放单号")
    name = fields.CharField(max_length=200, description="发放单名称")
    status = fields.CharField(
        max_length=20,
        default="Draft",
        description="Draft/Pending/Issued/Recalled",
    )
    remark = fields.TextField(null=True, description="备注")
    issued_at = fields.DatetimeField(null=True, description="发放时间")
    issued_by = fields.IntField(null=True, description="发放人ID")
    issued_by_name = fields.CharField(max_length=100, null=True, description="发放人姓名")
    recalled_at = fields.DatetimeField(null=True, description="收回时间")
    recalled_by = fields.IntField(null=True, description="收回人ID")
    recalled_by_name = fields.CharField(max_length=100, null=True, description="收回人姓名")
    recall_reason = fields.TextField(null=True, description="收回原因")
    deleted_at = fields.DatetimeField(null=True, description="软删除")
    created_by = fields.IntField(null=True, description="创建人ID")
    created_by_name = fields.CharField(max_length=100, null=True, description="创建人姓名")
    updated_by = fields.IntField(null=True, description="更新人ID")
    updated_by_name = fields.CharField(max_length=100, null=True, description="更新人姓名")


class DrawingDistributionLine(BaseModel):
    class Meta:
        table = "apps_master_data_drawing_distribution_lines"
        table_description = "基础数据管理 - 图档发放明细"
        indexes = [
            ("tenant_id",),
            ("distribution_id",),
            ("drawing_id",),
        ]

    id = fields.IntField(pk=True, description="主键ID")
    distribution_id = fields.IntField(description="发放单ID")
    drawing_id = fields.IntField(description="图纸ID")
    drawing_uuid = fields.CharField(max_length=36, description="图纸UUID")
    drawing_code = fields.CharField(max_length=50, description="图号")
    drawing_name = fields.CharField(max_length=200, description="图纸名称")
    drawing_revision = fields.CharField(max_length=20, description="修订版")
